prow/git/executor.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Executor knows how to execute Git commands
+type Executor interface {
+	Run(args ...string) ([]byte, error)
+}
+
+func NewCensoringExecutor(dir string, censor func(content []byte) []byte, logger *logrus.Entry) (Executor, error) {
+	g, err := exec.LookPath("git")
+	if err != nil {
+		return nil, err
+	}
+	return &censoringExecutor{
+		logger: logger,
+		dir:    dir,
```

Comment:
And generally: What about instead of building something completely new adding the option to get a `*Repo` from the `Client` by passing in a dir and possibly some meta info? That way we have the same functionality regardless of how the repo is fetched

Reply:
Cloning is not part of this client. They are separate things. I do not want to reuse the current client, since it does something _very_ specific and while they both manipulate `git`, they do so in very different ways. I think we would do best to have a pointed client with nice UX for each set of users.

Reply:
But we will end up re-implementing whats in the current client or factor it out and make both clients have thin wrappers around the actual logic.

What is your objection with a `func(c *Client) UseExistingRepository(path string)(*Repo, error)`?

Reply:
I don't see any overlap whatsoever. What are you worried about reimplemeting?

Reply:
(I also _specifically_ do not like the  `Repo` type since it's entirely untested.)

Reply:
> I  don't see any overlap whatsoever. What are you worried about reimplemeting?

Well for example there is already a `Push` implementation on the repo type

> (I also specifically do not like the Repo type since it's entirely untested.)

Fair point but isn't the solution to that to write tests instead of creating something new?